src/box/iproto.cc

```diff
@@ -557,6 +589,54 @@ struct iproto_connection
 	struct iproto_thread *iproto_thread;
 };
 
+static struct fiber_cond alive_connection_exists;
+
+void
+iproto_wait_close_all_connections(void)
+{
+	if (iproto_connection_count() > 0)
+		fiber_cond_wait(&alive_connection_exists);
+}
+
+void
+check_no_active_connections(struct iproto_thread *iproto_thread);
+
+void
+tx_check_no_active_connections(struct cmsg *msg)
+{
+	(void) msg;
+	if (iproto_connection_count() == 0)
+		fiber_cond_broadcast(&alive_connection_exists);
+}
+
+void
+net_end_check_connections(struct cmsg *msg)
+{
+	struct iproto_thread *iproto_thread =
+		container_of(msg, struct iproto_thread, checking_alive_connections_msg);
+	if (iproto_thread->checking_alive_connections_status == CHECKING_REQUEST)
+		check_no_active_connections(iproto_thread);
+	else
+		iproto_thread->checking_alive_connections_status = CHECKING_NOT_START;
+}
+
+void
+check_no_active_connections(struct iproto_thread *iproto_thread)
+{
+	if (iproto_thread->checking_alive_connections_status == CHECKING_PROCESS ||
+	    iproto_thread->checking_alive_connections_status == CHECKING_REQUEST) {
+		iproto_thread->checking_alive_connections_status = CHECKING_REQUEST;
```

Comment:
Here is some error as i see. If this `if` became` true` this function stop it's work at all.
For example client creates several connections
```lua
table = {}
for i = 1, 1000 do
    local graceful_con = net_box.connect("/tmp/1")
    graceful_con.space._session_settings:update('graceful_shutdown', {{'=', 'value', true}})
    net_box.set_shutdown_handler(function() graceful_con:shutdown({is_async=true}) end)
    table[#table + 1] = graceful_con
end
```
And then closes them in loop:
```lua
for i = 1, #table do
    table[i]:close()
end
```

See:
(1) iproto_thread->checking_alive_connections_status = CHECKING_PROCESS;
(2) Message send to tx thread and `tx_check_no_active_connections` called
(3) At this moment next `check_no_active_connections` called and
     iproto_thread->checking_alive_connections_status = CHECKING_REQUEST;
(4) `net_end_check_connections` called  with
       iproto_thread->checking_alive_connections_status == CHECKING_REQUEST
       and returns without changing `iproto_thread->checking_alive_connections_status`
From this time all `check_no_active_connections` calls returns immediately.

Reply:
Added changing of `checking_alive_connections_status` to `CHECK_NOT_START` before calling `check_no_active_connections`.